packages/core/database/src/migrations/internal-migrations/5.0.0-02-document-id.ts

```diff
@@ -0,0 +1,174 @@
+    const res = await knex.raw(
+      `
+    SELECT :tableName:.id as id, string_agg(DISTINCT :inverseJoinColumn:::character varying, ',') as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :tableName:.id, :joinColumn:
+    LIMIT 1;
+  `,
+      params
+    );
+
+    return res.rows;
+  },
+  async mysql(knex: Knex, params: Params) {
+    const [res] = await knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :tableName:.id, :joinColumn:
+    LIMIT 1;
+  `,
+      params
+    );
+
+    return res;
+  },
+  async sqlite(knex: Knex, params: Params) {
+    return knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :joinColumn:
+    LIMIT 1;
+    `,
+      params
+    );
+  },
+};
+
+const getNextIdsToCreateDocumentId = async (
+  db: Database,
+  knex: Knex,
+  {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  }: {
+    joinColumn: string;
+    inverseJoinColumn: string;
+    tableName: string;
+    joinTableName: string;
+  }
+): Promise<number[]> => {
+  const res = await QUERIES[db.dialect.client as keyof typeof QUERIES](knex, {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  });
+
+  if (res.length > 0) {
+    const row = res[0];
+    const otherIds = row.other_ids
+      ? row.other_ids.split(',').map((v: string) => parseInt(v, 10))
+      : [];
+
+    return [row.id, ...otherIds];
+  }
+
+  return [];
+};
+
+// Migrate document ids for tables that have localizations
+const migrateDocumentIdsWithLocalizations = async (db: Database, knex: Knex, meta: Meta) => {
+  const singularName = meta.singularName.toLowerCase();
+  const joinColumn = identifiers.getJoinColumnAttributeIdName(singularName);
+  const inverseJoinColumn = identifiers.getInverseJoinColumnAttributeIdName(singularName);
+
+  let ids: number[];
+
+  do {
+    ids = await getNextIdsToCreateDocumentId(db, knex, {
+      joinColumn,
+      inverseJoinColumn,
+      tableName: meta.tableName,
+      joinTableName: identifiers.getJoinTableName(meta.tableName, `localizations`),
+    });
+
+    if (ids.length > 0) {
+      await knex(meta.tableName).update({ document_id: createId() }).whereIn('id', ids);
+    }
+  } while (ids.length > 0);
+};
+
+// Migrate document ids for tables that don't have localizations
+const migrationDocumentIds = async (db: Database, knex: Knex, meta: Meta) => {
+  let res;
+  do {
+    res = await knex.select('id').from(meta.tableName).whereNull('document_id').limit(1).first();
```

Comment:
Can we do that in a single query 🤔 

```sql
UPDATE xxx
SET document_id = "new-id"
WHERE id in (SELECT id FROM xxx WHERE document_id IS NULL LIMIT 1)
```

And then we check the response has had an effect (count of affected rows)

All in All for this query we should do a count, generates a temp table with x document_id in one query & then do a 1:1 update in a single query

Reply:
Done! Let me know what you think
[`4d49271` (#19514)](https://github.com/strapi/strapi/pull/19514/commits/4d49271d01026ba3662dd45f24ee71cdb9538033)